### PR TITLE
feat: new `adminSecret` option

### LINKED ISSUE #193: Using admin secret
- methods for setting and clearing admin secret
- if admin secret is set, it should be sent along in the headers for all requests (`x-hasura-admin-secret`)

This is useful when sending GraphQL requests, managing files, signing up users, and sending function calls.

It should also be easy to toggle between making requests using the admin secret or not.

### PR DESCRIPTION
**Pull request by plmercereau**
close #193

### CODE REVIEW

File: packages/hasura-storage-js/src/hasura-storage-client.ts [hunk 1]
```diff
@@ -9,16 +9,28 @@ import {
 } from './utils/types'
 import { HasuraStorageApi } from './hasura-storage-api'
 
+interface NhostStorageConstructorParams {
+  /**
+   * Hasura storage endpoint.
```
Comment (elitan) [inline packages/hasura-storage-js/src/hasura-storage-client.ts:14]: Storage endpoint

**Review by elitan — CHANGES_REQUESTED** [review]
Nice! Small copy changes.

In the future, I think we should add methods for interacting with users as admin too. Like:

```
nhost.auth.admin.createUser()
nhost.auth.admin.updateUser()
nhost.auth.admin.deleteUser()
```

**Comment by plmercereau** [comment]
> Nice! Small copy changes.
> 
> In the future, I think we should add methods for interacting with users as admin too. Like:
> 
> ```
> nhost.auth.admin.createUser()
> nhost.auth.admin.updateUser()
> nhost.auth.admin.deleteUser()
> ```

Yes, we should think of a way to implement this without increasing the complexity of our code nor bypassing the existing workflows (for instance email verification)

**Review by elitan — APPROVED** [review]

**Comment by st3phan** [comment]
@plmercereau I tried this new `adminSecret` option in my functions set up and it works.
But in my use case when using `nhost.graphql.request` I sometimes want to overwrite the authorization headers with client headers that are forwarded to the webhook by hasura.
When I do it seems that the authorization headers are ignored.

In code:
```ts
export const nhost = new NhostClient({
  backendUrl: process.env.NHOST_BACKEND_URL,
  adminSecret: process.env.NHOST_ADMIN_SECRET
});

const { data, error } = await nhost.graphql.request(
  DOCUMENT,
  {
    headers: {
      authorization: req.headers.authorization, // Looks like this is ignored because of the `adminSecret` that is set
      'x-hasura-user-id': req.session_variables['x-hasura-user-id'],
    },
  },
);
```

**Comment by plmercereau** [comment]
> @plmercereau I tried this new `adminSecret` option in my functions set up and it works. But in my use case when using `nhost.graphql.request` I sometimes want to overwrite the authorization headers with client headers that are forwarded to a webhook by hasura. When I do it seems that the authorization headers are ignored.
> 
> In code:
> 
> ```ts
> export const nhost = new NhostClient({
>   backendUrl: process.env.NHOST_BACKEND_URL,
>   adminSecret: process.env.NHOST_ADMIN_SECRET // I want to use the adminSecret for all requests except when I set custom headers
> });
> 
> const { data, error } = await nhost.graphql.request(
>   DOCUMENT,
>   {
>     headers: {
>       authorization: req.headers.authorization, // Looks like this is ignored because of the `adminSecret` that is set
>       'x-hasura-user-id': req.session_variables['x-hasura-user-id'],
>     },
>   },
> );
> ```

I'm repopening the issue and will take some time for reproduction and analysis.
Will keep you posted

Thanks mate

**Comment by st3phan** [comment]
@plmercereau any update on this?

**Comment by GavanWilhite** [comment]
@st3phan @plmercereau End up figuring anything out here?

**Pull request merged**